### PR TITLE
Add cuDNN flash attention sequence packing

### PR DESCRIPTION
**Pull request by Cjkkkk**
* add cudnn flash attention sequence packing support where multiple batches(segments) can be packed into one batch. 
* add two extra arguments **q_offsets** and **kv_offsets**. Offsets tensor specify the starting pos of each segment inside one batch. **q_seqlen** and **kv_seqlen** will be also required to specify the actual sequence length of each segment in case of paddings.
*  cudnn accepts **q_offsets** of shape [S] where S is the number of segments. Since S can change at runtime from batch to batch, we design it to have shape [B, M] where B is the number of batches and M is the maximum number of segments. Therefore static allocation can be used and all non zero entries of q_offsets will be shift to left and form a shape [S] tensor before passing to cuDNN.
* the unit test is comparing using segment_id to generate segment_mask to cuDNN.

Related XLA PR: https://github.com/openxla/xla/pull/20861

### CODE REVIEW

**Comment by Cjkkkk** [comment]
@superbobry Hi Sergei, could you help review this PR? Thanks!

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 1]
```diff
@@ -1648,7 +1729,21 @@ def dot_product_attention(
     bias: Bias to be added to logits with a shape of BNTS.
     mask: Mask used to filter out logits with a shape of BNTS.
     q_seqlen: Non padded sequence length of Queries with a shape of B.
+      If q_offsets is set, q_seqlen should have shape [B,M] where M is the
+      maximum number of segments per batch. Fill the invalid entries with -1.
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:1733]: "invalid" looks a bit out of place here. Would it be more clear to say that -1 is the padding value?

Please update other occurrences if you update this one.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 2]
```diff
@@ -1648,7 +1729,21 @@ def dot_product_attention(
...
     kv_seqlen: Non padded sequence length of Keys and Values with a shape of B.
+      If kv_offsets is set, kv_seqlen should have shape [B,M] where M is the
+      maximum number of segments per batch. Fill the invalid entries with -1.
+    q_offsets: offset of each segment packed in Queries with a shape of [B,M+1]
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:1737]: Should it be "query" instead of "Queries", since you are probably referring to the argument?

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 3]
```diff
@@ -1648,7 +1729,21 @@ def dot_product_attention(
...
+      where M is the maximum number of segments per batch. Fill the invalid
+      entries with -1. E.g, if 2 batches has 3 and 2 segments respectively,
+      each segment has size 1, q_offsets = [[0,1,2,-1], [0,1,-1,-1]].
+      q_seqlen should be set to indicate the size of each segment.
+    kv_offsets: offset of each segment packed in Keys with a shape of [B,M+1]
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:1742]: As above, use "key" and "value" if you want to refer to the corresponding arguments.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 4]
```diff
@@ -1724,8 +1821,12 @@ def dot_product_attention(
       q_seqlen = jnp.zeros(0, dtype=query.dtype)
     if kv_seqlen is None:
       kv_seqlen = jnp.zeros(0, dtype=query.dtype)
+    if q_offsets is None:
+      q_offsets = jnp.zeros(0, dtype=query.dtype)
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:1825]: Do we have to do this? Can we pass `None` instead?

Also OOC why `zeros` and not `jnp.full(-1, ...)`?

File: tests/fused_attention_stablehlo_test.py [hunk 5]
```diff
@@ -612,6 +614,127 @@ def test_sdpa_large_head_size(self):
     self.assertArraysAllClose(grads_ref[1], grads_ans[1])
     self.assertArraysAllClose(grads_ref[2], grads_ans[2])
 
+  @jtu.run_on_devices("cuda")
+  def test_sdpa_packed_layout(self):
```
Comment (superbobry) [inline tests/fused_attention_stablehlo_test.py:618]: This test probably needs a cuDNN version check.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 6]
```diff
@@ -378,78 +386,131 @@ def check_compute_capability(capability):
...
+      activation, fwd_output, grad_output, scale=scale, seed=seed,
       dropout_rate=dropout_rate, variadic_args=variadic_args,
       mask_type=mask_type, layout=layout,
       sliding_window_length=sliding_window_length
   )
-  grads = (*grads,) + (None,) * (6 - len(grads))
+  grads = (*grads,) + (None,) * (8 - len(grads))
   return grads
 
+def _fix_seqlen_offsets(q_seqlen, kv_seqlen, q_offsets, kv_offsets, query, key):
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:436]: This function needs a docstring explaining what "fix" means.

**Review by superbobry — APPROVED** [review]

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 4 above)
Comment (Cjkkkk) [reply jax/_src/cudnn/fused_attention_stablehlo.py:1825]: I think None is not a valid jax type? So what I do is to create a dummy tensor as a WAR. Another thing is that I want to add this tensor in nondiff_argnums but I think jax expects array to not be not included but return None in bwd to achieve the same. Do we have to do this or is a better way to achieve the same behavior?

**Comment by Cjkkkk** [comment]
@superbobry Thanks for reviewing this. I will address the comments by EOD, for the tests failure, I think I can address the lint_and_typecheck one but I think other failure aren't related? For example, CI - Bazel CPU tests (RBE).

**Comment by superbobry** [comment]
Thanks, please address in lint_and_typecheck.

**Comment by Cjkkkk** [comment]
@superbobry I addressed lint_and_typecheck could you take another look? Thanks!

**Review by superbobry — APPROVED** [review]

**Comment by superbobry** [comment]
It looks like there is still a type error remaining?

```
jax/_src/cudnn/fused_attention_stablehlo.py:1792: error: Missing positional arguments "kv_offsets", "layout" in call to "check_layout"  [call-arg]
```
